src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java

```diff
@@ -181,10 +182,15 @@ private void doStart() throws Exception {
 
         @Override public void expand(EnvVars env) throws IOException, InterruptedException {
             for (Map.Entry<String,Secret> override : overrides.entrySet()) {
-                env.override(override.getKey(), override.getValue().getPlainText());
+                String keyOverride = override.getKey();
+                env.override(keyOverride, override.getValue().getPlainText());
+                sensitiveVars.add(keyOverride);
             }
         }
 
+        @Override public Set<String> getSensitiveVars() {
+            return Collections.unmodifiableSet(sensitiveVars);
```

Comment:
I think you could simplify things so you don't even have to add the `sensitiveVars` field:

```suggestion
            return Collections.unmodifiableSet(overrides.keySet());
```

I think it is normal or at least possible for `expand` to be called more than once on different instances of `EnvVars`, and the current code would try to update `sensitiveVars` every time `expand` is called, which would have no effect since the field is a `Set`, but is a bit redundant. 

Also, the current implementation means that `getSensitiveVars` returns an empty set until `expand` has been called at least once, which in practice might be fine but it seems like it could cause confusion at some point; using `overrides.getKeySet()` avoids this problem.

Reply:
regarding the deserialization problem, would it be enough to add `null`  checks whenever `sensitiveVars` is accessed?